### PR TITLE
[MOV] purchase,purchase_stock,stock_dropshipping: move _prepare_purchase_order_line to be usable without stock

### PR DESCRIPTION
**Pull request by svs-odoo**
**[MOV]** ***: _prepare_purchase_order_line**
> Move the method `_prepare_purchase_order_line` from 'stock.rule' to 'purchase.order.line'.
> The purpose is to reuse it in approvals_purchase.
> 
>task-2040710

### CODE REVIEW

File: addons/purchase/models/purchase.py [hunk 1]
```diff
@@ -755,3 +755,47 @@ def _prepare_account_move_line(self, move):
...
+
+        product_lang = product_id.with_context(
+            lang=partner.lang,
+            partner_id=partner.id,
+        )
+        name = product_lang.display_name
+        if product_lang.description_purchase:
+            name += '\n' + product_lang.description_purchase
+
+        date_planned = self.env['purchase.order.line']._get_date_planned(seller, po=po)
```
Comment (mreficent) [inline addons/purchase/models/purchase.py]:
```suggestion
        date_planned = self._get_date_planned(seller, po=po)
```
right?

File: addons/purchase/models/purchase.py [hunk 2]
```diff
@@ -755,3 +755,47 @@ def _prepare_account_move_line(self, move):
             'tax_ids': [(6, 0, self.taxes_id.ids)],
             'display_type': self.display_type,
         }
+
+    @api.model
+    def _prepare_purchase_order_line(self, product_id, product_qty, product_uom, company_id, values, po):
```
Comment (amoyaux) [inline addons/purchase/models/purchase.py]: pass supplier directly instead of values

File: addons/purchase_stock/models/purchase.py [hunk 3]
```diff
@@ -388,6 +388,16 @@ def _prepare_stock_moves(self, picking):
             res.append(template)
         return res
 
+    @api.model
+    def _prepare_purchase_order_line(self, product_id, product_qty, product_uom, company_id, values, po):
+        res = super()._prepare_purchase_order_line(product_id, product_qty, product_uom, company_id, values, po)
```
Comment (amoyaux) [inline addons/purchase_stock/models/purchase.py]: Do a separate function _prepare_purchase_order_line_from_procurement that call _prepare_purchase_order_line

File: addons/purchase_stock/models/purchase.py [hunk 4]
```diff
@@ -388,6 +388,17 @@ def _prepare_stock_moves(self, picking):
             res.append(template)
         return res
 
+    @api.model
+    def _prepare_purchase_order_line_from_procurement(self, product_id, product_qty, product_uom, company_id, values, po):
+        supplier = values.get('supplier')
+        res = super()._prepare_purchase_order_line(product_id, product_qty, product_uom, company_id, supplier, po)
```
Comment (sle-odoo) [inline addons/purchase_stock/models/purchase.py]: self. _prepare_purchase_order_line

**Comment by sle-odoo** [comment]
robodoo r+

**Pull request closed without merging**